File: app/src/main/java/br/com/stonesdk/sdkdemo/activities/BaseTransactionActivity.java
package br.com.stonesdk.sdkdemo.activities;

import android.app.Dialog;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.RadioGroup;
import android.widget.RelativeLayout;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import br.com.stonesdk.sdkdemo.R;
import stone.application.enums.Action;
import stone.application.enums.ErrorsEnum;
import stone.application.enums.InstalmentTransactionEnum;
import stone.application.enums.TypeOfTransactionEnum;
import stone.application.interfaces.StoneActionCallback;
import stone.database.transaction.TransactionObject;
import stone.providers.BaseTransactionProvider;
import stone.user.UserModel;
import stone.utils.Stone;

/**
 * Created by felipe on 05/03/18.
 */
public abstract class BaseTransactionActivity<T extends BaseTransactionProvider> extends AppCompatActivity implements StoneActionCallback {
    private BaseTransactionProvider transactionProvider;
    protected final TransactionObject transactionObject = new TransactionObject();
    RadioGroup transactionTypeRadioGroup;
    Spinner installmentsSpinner;
    Spinner stoneCodeSpinner;
    TextView installmentsTextView;
    CheckBox captureTransactionCheckBox;
    EditText amountEditText;
    TextView logTextView;
    Button sendTransactionButton;
    Button cancelTransactionButton;
    Dialog builder;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_transaction);
        transactionTypeRadioGroup = findViewById(R.id.transactionTypeRadioGroup);
        installmentsTextView = findViewById(R.id.installmentsTextView);
        installmentsSpinner = findViewById(R.id.installmentsSpinner);
        stoneCodeSpinner = findViewById(R.id.stoneCodeSpinner);
        captureTransactionCheckBox = findViewById(R.id.captureTransactionCheckBox);
        amountEditText = findViewById(R.id.amountEditText);
        logTextView = findViewById(R.id.logTextView);
        sendTransactionButton = findViewById(R.id.sendTransactionButton);
        cancelTransactionButton = findViewById(R.id.cancelTransactionButton);
        spinnerAction();
        radioGroupClick();
        sendTransactionButton.setOnClickListener(v -> initTransaction());
        cancelTransactionButton.setOnClickListener(v -> {
            if (transactionProvider != null) {
                new Thread(new Runnable() {
                    @Override
                    public void run() {
                        transactionProvider.abortPayment();
                    }
                }).start();
            }
        });
        builder = new Dialog(this);
        builder.requestWindowFeature(Window.FEATURE_NO_TITLE);
        builder.getWindow().setBackgroundDrawable(
                new ColorDrawable(android.graphics.Color.TRANSPARENT));
    }

    private void radioGroupClick() {
        transactionTypeRadioGroup.setOnCheckedChangeListener((group, checkedId) -> {
            if (checkedId == R.id.radioPix || checkedId == R.id.radioDebit || checkedId == R.id.radioVoucher) {
                installmentsTextView.setVisibility(View.GONE);
                installmentsSpinner.setVisibility(View.GONE);
            } else if (checkedId == R.id.radioCredit) {
                installmentsTextView.setVisibility(View.VISIBLE);
                installmentsSpinner.setVisibility(View.VISIBLE);
            }
        });
    }

    private void spinnerAction() {
        ArrayAdapter<CharSequence> adapter = ArrayAdapter.createFromResource(this, R.array.installments_array, android.R.layout.simple_spinner_item);
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        installmentsSpinner.setAdapter(adapter);
        ArrayAdapter<String> stoneCodeAdapter = new ArrayAdapter<>(this, android.R.layout.simple_list_item_1, android.R.id.text1);
        stoneCodeAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        if (Stone.sessionApplication != null && Stone.sessionApplication.getUserModelList() != null) {
            for (UserModel userModel : Stone.sessionApplication.getUserModelList()) {
                stoneCodeAdapter.add(userModel.getStoneCode());
            }
        }
        stoneCodeSpinner.setAdapter(stoneCodeAdapter);
    }

    public void initTransaction() {
        InstalmentTransactionEnum installmentsEnum = InstalmentTransactionEnum.getAt(installmentsSpinner.getSelectedItemPosition());
        // Informa a quantidade de parcelas.
        transactionObject.setInstalmentTransaction(InstalmentTransactionEnum.getAt(installmentsSpinner.getSelectedItemPosition()));
        // Verifica a forma de pagamento selecionada.
        TypeOfTransactionEnum transactionType;
        int checkedRadioButtonId = transactionTypeRadioGroup.getCheckedRadioButtonId();
        if (checkedRadioButtonId == R.id.radioCredit) {
            transactionType = TypeOfTransactionEnum.CREDIT;
        } else if (checkedRadioButtonId == R.id.radioDebit) {
            transactionType = TypeOfTransactionEnum.DEBIT;
        } else if (checkedRadioButtonId == R.id.radioVoucher) {
            transactionType = TypeOfTransactionEnum.VOUCHER;
        } else if (checkedRadioButtonId == R.id.radioPix) {
            transactionType = TypeOfTransactionEnum.PIX;
        } else {
            transactionType = TypeOfTransactionEnum.CREDIT;
        }
        transactionObject.setInitiatorTransactionKey(null);
        transactionObject.setTypeOfTransaction(transactionType);
        transactionObject.setCapture(captureTransactionCheckBox.isChecked());
        transactionObject.setAmount(amountEditText.getText().toString());
        transactionProvider = buildTransactionProvider();
        transactionProvider.setConnectionCallback(this);
        new Thread(new Runnable() {
            @Override
            public void run() {
                transactionProvider.execute();
            }
        }).start();
    }

    protected String getAuthorizationMessage() {
        return transactionProvider.getMessageFromAuthorize();
    }

    protected abstract T buildTransactionProvider();

    protected boolean providerHasErrorEnum(ErrorsEnum errorsEnum) {
        return transactionProvider.theListHasError(errorsEnum);
    }

    @Override
    public void onError() {
        runOnUiThread(() -> Toast.makeText(BaseTransactionActivity.this, "Erro: " + transactionProvider.getListOfErrors(), Toast.LENGTH_SHORT).show());
    }

    @Override
    public void onStatusChanged(final Action action) {
        runOnUiThread(() -> logTextView.append(action.name() + "\n"));
        if (action == Action.TRANSACTION_WAITING_QRCODE_SCAN) {
            ImageView imageView = new ImageView(this);
            imageView.setImageBitmap(transactionObject.getQRCode());

            runOnUiThread(() -> {
                builder.addContentView(imageView, new RelativeLayout.LayoutParams(
                        ViewGroup.LayoutParams.MATCH_PARENT,
                        ViewGroup.LayoutParams.MATCH_PARENT));
                builder.show();
            });
        } else {
            runOnUiThread(() -> builder.dismiss());
        }
    }

    protected BaseTransactionProvider getTransactionProvider() {
        return transactionProvider;
    }

    protected UserModel getSelectedUserModel() {
        if (Stone.sessionApplication != null && Stone.sessionApplication.getUserModelList() != null
                && !Stone.sessionApplication.getUserModelList().isEmpty()) {
            return Stone.getUserModel(stoneCodeSpinner.getSelectedItemPosition());
        }
        return null;
    }
}

File: app/src/main/java/br/com/stonesdk/sdkdemo/activities/transaction/BaseTransactionActivity.kt
package br.com.stonesdk.sdkdemo.activities.transaction

//import stone.application.enums.Action
//import stone.application.enums.ErrorsEnum
//import stone.application.enums.InstalmentTransactionEnum
//import stone.application.enums.TypeOfTransactionEnum
//import stone.application.interfaces.StoneActionCallback
//import stone.database.transaction.TransactionObject
//import stone.providers.BaseTransactionProvider
//import stone.user.UserModel
//import stone.utils.Stone


//abstract class BaseTransactionActivity<T : BaseTransactionProvider?> : AppCompatActivity(),
//    StoneActionCallback {
//    private lateinit var binding: ActivityTransactionBinding
//
//    protected var transactionProvider: BaseTransactionProvider? = null
//        private set
//
//    @JvmField
//    protected val transactionObject: TransactionObject = TransactionObject()
//
//    private var builder: Dialog? = null
//
//    public override fun onCreate(savedInstanceState: Bundle?) {
//
//        super.onCreate(savedInstanceState)
//        if (FeatureFlag.composeRefactorEnabled){
//            setContent {
//                MaterialTheme {
//                    TransactionScreen()
//                }
//            }
//        } else {
//            onCreateConfig()
//        }
//    }
//
//    fun onCreateConfig(){
//        binding = ActivityTransactionBinding.inflate(layoutInflater)
//        setContentView(binding.root)
//
//        spinnerAction()
//        radioGroupClick()
//        binding.apply {
//            sendTransactionButton.setOnClickListener { initTransaction() }
//            cancelTransactionButton.setOnClickListener { transactionProvider?.abortPayment() }
//        }
//
//        builder = Dialog(this).apply {
//            requestWindowFeature(Window.FEATURE_NO_TITLE)
//            window?.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
//        }
//    }
//
//    private fun radioGroupClick() = with(binding) {
//        transactionTypeRadioGroup.setOnCheckedChangeListener { _, checkedId ->
//            when (checkedId) {
//                R.id.radioPix, R.id.radioDebit, R.id.radioVoucher -> {
//                    installmentsTextView.visibility = View.GONE
//                    installmentsSpinner.visibility = View.GONE
//                }
//
//                R.id.radioCredit -> {
//                    installmentsTextView.visibility = View.VISIBLE
//                    installmentsSpinner.visibility = View.VISIBLE
//                }
//            }
//        }
//    }
//
//    private fun spinnerAction() = with(binding) {
//        val adapter = ArrayAdapter.createFromResource(
//            this@BaseTransactionActivity,
//            R.array.installments_array,
//            android.R.layout.simple_spinner_item
//        )
//        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
//        installmentsSpinner.adapter = adapter
//
//        val stoneCodeAdapter = ArrayAdapter<String>(
//            this@BaseTransactionActivity,
//            android.R.layout.simple_list_item_1, android.R.id.text1
//        )
//        stoneCodeAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
//        for (userModel in Stone.sessionApplication.userModelList) {
//            stoneCodeAdapter.add(userModel.stoneCode)
//        }
//        stoneCodeSpinner.adapter = stoneCodeAdapter
//    }
//
//    private fun initTransaction() = with(binding) {
//        val installmentsEnum = InstalmentTransactionEnum.getAt(
//            installmentsSpinner.selectedItemPosition
//        )
//
//        // Informa a quantidade de parcelas.
//        transactionObject.instalmentTransaction = InstalmentTransactionEnum.getAt(
//            installmentsSpinner.selectedItemPosition
//        )
//
//        // Verifica a forma de pagamento selecionada.
//        val transactionType = when (transactionTypeRadioGroup.checkedRadioButtonId) {
//            R.id.radioCredit -> TypeOfTransactionEnum.CREDIT
//            R.id.radioDebit -> TypeOfTransactionEnum.DEBIT
//            R.id.radioVoucher -> TypeOfTransactionEnum.VOUCHER
//            R.id.radioPix -> TypeOfTransactionEnum.PIX
//            else -> TypeOfTransactionEnum.CREDIT
//        }
//        transactionObject.initiatorTransactionKey = null
//        transactionObject.typeOfTransaction = transactionType
//        transactionObject.isCapture = captureTransactionCheckBox.isChecked
//        transactionObject.amount = amountEditText.text.toString()
//
//        transactionProvider = buildTransactionProvider()
//        transactionProvider?.setConnectionCallback(this@BaseTransactionActivity)
//        transactionProvider?.execute()
//    }
//
//    protected val authorizationMessage: String?
//        get() = transactionProvider?.messageFromAuthorize
//
//    protected abstract fun buildTransactionProvider(): T
//
//    protected fun providerHasErrorEnum(errorsEnum: ErrorsEnum?): Boolean {
//        return transactionProvider?.theListHasError(errorsEnum) ?: false
//    }
//
//    override fun onError() {
//        runOnUiThread {
//            Toast.makeText(
//                this@BaseTransactionActivity,
//                "Erro: " + transactionProvider?.listOfErrors,
//                Toast.LENGTH_SHORT
//            ).show()
//        }
//    }
//
//    override fun onStatusChanged(action: Action) {
//        runOnUiThread { binding.logTextView.append(action.name + "\n") }
//
//        if (action == Action.TRANSACTION_WAITING_QRCODE_SCAN) {
//            val imageView = ImageView(this)
//            imageView.setImageBitmap(transactionObject.qrCode)
//
//            runOnUiThread {
//                builder?.addContentView(
//                    imageView, RelativeLayout.LayoutParams(
//                        ViewGroup.LayoutParams.MATCH_PARENT,
//                        ViewGroup.LayoutParams.MATCH_PARENT
//                    )
//                )
//                builder?.show()
//            }
//        } else {
//            runOnUiThread { builder?.dismiss() }
//        }
//    }
//
//    protected val selectedUserModel: UserModel
//        get() = Stone.getUserModel(binding.stoneCodeSpinner.selectedItemPosition)
//}
